design/StableRegions-20230209.md

```diff
@@ -0,0 +1,222 @@
+
+The `_release` operation is *not* part of the user-facing API nor part of the MVP,
+but supporting it is important because it means we can transition quickly to an integration
+with the ambient Motoko GC if we can support it.
+
+## Internal footprint
+
+The state of the allocator is stored in a combination of stable memory and heap memory.
+
+The stable memory state is sufficient to reconstitute the heap objects
+(`rebuild` operation, described in a subsection below).  That means
+that even if the heap is lost (like during an upgrade that fails) the
+stable memory state can fully describe the parts that should be
+rebuilt into the heap when the upgrade succeeds.
+
+### stable memory fields
+
+ - total allocated blocks, `Nat16`, max value is `32768`.
+ - total allocated regions, `Nat16`, max value is `32767` (one region is reserved for "no region").
+ - The `block-region` table (fixed size, about 131kb).
+ - The `region-blocks` table (fixed size, about 524kb).
+
+### representation of values of type `Region`
+
+ - We use the region ID, a `Nat16`, to represent the region as a value.
+ - To compute a load or store access, this requires consulting stable memory for (almost all) meta data.
+ - We rely on the cost of these stable memory accesses going down in the near future to support this design decision.
+
+### block-region table
+
+ - purpose:
+   - relate a block ID ("page block ID") to its region (if any) and its position in that region (see `rebuild`).
+
+   - NB: The organization of this table is not useful for efficient
+     access calculations of load or store (they require a linear
+     search that would be prohibitively slow). Rather, the
+     `region-blocks` table is designed for that purpose.  OTOH, this
+     table is suitable to do a "batch" rebuild of the dynamic heap-allocated vectors
+     in that table, which get lost during upgrades.
+
+ - 32768 entries (statically sized).
+ - 4 bytes per entry.
+ - entry type = `BlockRegion { region : Nat16, position : Nat16 }`
+ - the location of each entry gives its corresponding block ID.
+
+### region-blocks table
+
+ - purpose: 
+   - relate a region ID to its vector of block IDs, to compute an access efficiently (load or store).
+
+   - NB: The organization of this table is meant to support O(1) load
+     and store operations, but in so doing, it needs to use a
+     dynamically-sized vector for each region.  There is no a priori
+     way to know how to allocate these, and if we naively preallocate
+     them to each be the potential "maximal region" (with all blocks
+     allocated to it), the resulting preallocated table requires
+     gigabytes of space, and thus is prohibitively large.
+
+     At the same time, having dynamically-sized stable vectors is kind
+     of the point of regions (each is such a vector, in a sense), and
+     so requiring these first to implement regions' meta data seems
+     potentially circular in concept, if not also very complex.
```

Comment:
I'm pretty sure you can just use a blob, the region representation is shared and has identity, the blob will uniquely owned by the region.
The GC ignores the blob contents, so you store offsets (not object references) into stablememory or block ids from which you can compute the offset.

Reply:
When we move to a world of collected regions, the GC may need to traverse the valid blob entries to mark the blocks as live, but only then.